### PR TITLE
Upgrade to Angular 11

### LINKED ISSUE #1289: Upgrade to Angular 12 (or 11)
Currently our codebase is on Angular 10.  As [Angular 10 LTS support ends on Dec 24](https://angular.io/guide/releases), we should look at upgrading to either Angular 11 or 12 in the 7.1 release.

If an upgrade to Angular 12 would prove difficult, simply getting us on Angular 11 is perfectly OK for 7.1.  However, if upgrading to Angular 12 is not much different than an Angular 11 upgrade, we may want to do both at once.

Use the Angular Upgrade guide for analysis: https://update.angular.io/

### PR DESCRIPTION
**Pull request by tdonohue**
## References
* Fixes #1289 by upgrading us to Angular 11.  (The Angular 12 upgrade can be rescheduled for 7.3 or 7.4 in a separate ticket.)
* Includes the minor changes in #1400 

## Description
This PR uses the automated upgrade process provided by Angular CLI to upgrade us from Angular v10.2 to v11.  The upgrade has proven to be easier than expected, as most all code changes were automated.

This upgrade was run by doing the following:
1. Ran `npx @angular/cli@11 update @angular/core@11 @angular/cli@11 --force`. See commit 9f1a017b56dc94957bc78ce779e939e0ff2c7d25. All those changes were automated.
    * Ran `ng update` to get a sense of which modules needed updating. Then updated those needing it...
2. Ran `ng update @angular/cdk@11`. See commit 0d031d0b252e7eb314e0a23a599525aeefc195c3
3. Ran `ng update @ngrx/store@11`. See commit 8fedb2c177523893facd83f21fd36f85eb2ed029
    * **SKIPPED** the "Automated migration of old `@Effect` annotations to new `createEffect()`", see https://ngrx.io/guide/migration/v11#ngrxeffects.  
        * In v11, `@Effect` was deprecated. So, we MUST run this migration before or during our upgrade to Angular 12.  
        * The provided automatic script _mostly worked_ but resulted in a few compilation errors in our codebase. I think an expert on `ngrx` might be able to solve those compilation issues rather quickly, but I wasn't able to easily understand how to refactor our existing code.
4. Ran `ng update @nguniversal/express-engine@11`.  See commit 2d633d79dc73a665c066da56c5e2fe4e1bbbc9c3
5. Ran `ng update ng-mocks@11`.  See commit be289617e127500eab3cccc5cd3e863967f350cd
6. Ran `ng update @ng-bootstrap/ng-bootstrap@9 --force`.  See commit ab3fa889130835f27fdfad7696239f6af2c11375
    * **REQUIRED MANUAL CHANGES:**  See 10c342483aa78d69279de2e4e040ae2ca1223254  I had to refactor some existing templates to no longer use `ngbTabSet` (which is removed), and instead use `ngbNav`.  See https://ng-bootstrap.github.io/#/components/tabset/api#NgbTabContent
    * Also had to manually fix some tests in commit d2cc18476382bc41dc5673d786030df43379bab9
7. Ran `ng update @ng-dynamic-forms/core@13 @ng-dynamic-forms/ui-ng-bootstrap@13 --force`.  See commit d926a24088e23db04e4e436ecbf4b5557ef31960
8. Ran `ng update ngx-sortablejs` to update to latest version . See commit 597e5396f9724f7435b75453a98d98978a53bd99
9. Finally, because my local Visual Studio Code had issues finding the correct version of Typescript, I added the final commit 466aaaa0d57bbfb98b1fb34e347b28f717db2cbd This one may be unrelated to Angular 11, and might be a local setup issue...but it seemed harmless to me.

## Instructions for Reviewers
* Test the application. I've tested and have found no issues in the upgrade so far.
* Review the **manual code changes**.  Especially review 10c342483aa78d69279de2e4e040ae2ca1223254 and d2cc18476382bc41dc5673d786030df43379bab9
* Verify all existing tests succeed.

### CODE REVIEW

**Comment by tdonohue** [comment]
@artlowel and @atarix83 : If you can find time to review this immediately (sometime this week), I'd appreciate it.  Last Friday I decided to investigate the complexity of simply upgrading to Angular 11....and as I was doing so, I realized that the upgrade is mostly "automated". Therefore, very little of this PR was code written by me.  I've been testing the UI on Angular 11, and so far have not seen any issues.

So, we may be able to quickly upgrade to Angular 11 if one or both of you can find time to help me test this out.

**Review by atarix83 — DISMISSED** [review]
thanks @tdonohue for this PR

I gave a review to the code and had quick tests on the application.
Generally it looks very good and Ididn't find any issues, I am only concerned about the incorrect peer dependencies.
I've seen that not all can be fixed but I'd ask if you can verify those and resolve as much as possible.

I'll made further tests on the application

**Comment by tdonohue** [comment]
@atarix83 : I believe almost all of those "incorrect peer dependencies" are unsolvable at this time.  I'll give them a second look later today, but I did review them initially and tried to solve many.  But, I found that unfortunately most of these dependencies have not seen updated releases in some time.  

Nonetheless, I'll give it a second look to see if there's any more that can be cleaned up immediately...others may need to wait or be treated as a separate ticket.

**Comment by wwelling** [comment]
Out of the unmet Angular peer dependencies below, `ngx-sortablejs` can be upgraded to `^11.1.0`. The others will require upgrading Angular dependencies further.

```
npm WARN @kolkov/ngx-gallery@1.2.3 requires a peer of tslib@^1.9.3 but none is installed. You must install peer dependencies yourself.
npm WARN @ng-dynamic-forms/ui-ng-bootstrap@13.0.0 requires a peer of @ng-bootstrap/ng-bootstrap@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @nicky-lenaers/ngx-scroll-to@9.0.0 requires a peer of @angular/common@^8.0.0 || ^9.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @nicky-lenaers/ngx-scroll-to@9.0.0 requires a peer of @angular/core@^8.0.0 || ^9.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @nicky-lenaers/ngx-scroll-to@9.0.0 requires a peer of tslib@^1.10.0 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.3.1 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
npm WARN bootstrap@4.3.1 requires a peer of popper.js@^1.14.7 but none is installed. You must install peer dependencies yourself.
npm WARN ngx-sortablejs@10.1.0 requires a peer of @angular/common@~10.2.0 but none is installed. You must install peer dependencies yourself.
npm WARN ngx-sortablejs@10.1.0 requires a peer of @angular/core@~10.2.0 but none is installed. You must install peer dependencies yourself.
npm WARN postcss-cli@8.3.1 requires a peer of postcss@^8.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN postcss-reporter@7.0.4 requires a peer of postcss@^8.1.0 but none is installed. You must install peer dependencies yourself.
npm WARN @angular-devkit/build-angular@0.1002.3 requires a peer of @angular/compiler-cli@^10.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @ngtools/webpack@10.2.3 requires a peer of @angular/compiler-cli@^10.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @ngtools/webpack@10.2.0 requires a peer of @angular/compiler-cli@^10.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN karma-jasmine-html-reporter@1.7.0 requires a peer of jasmine-core@>=3.8 but none is installed. You must install peer dependencies yourself.
npm WARN cssnano-utils@2.0.1 requires a peer of postcss@^8.2.15 but none is installed. You must install peer dependencies yourself.
```

**Comment by tdonohue** [comment]
@wwelling : Thanks for your analysis! That's helpful.

@atarix83 : I upgraded `ngx-sortablejs` to the latest version. That appears to be all we can do in this PR.

**Review by wwelling — APPROVED** [review]

**Comment by tdonohue** [comment]
I just did a rebase on `main` after #1412 was merged...as that PR resulted in some minor merge conflicts with this one in both `package.json` and `yarn.lock`.  Those merge conflicts have now all been resolved.

**Comment by wwelling** [comment]
@tdonohue @atarix83 @artlowel is there any chance we can raise the priority of this review? The sooner we can have upgraded dependencies to branch from the better.

**Comment by tdonohue** [comment]
@wwelling : I agree, I'd like to see this moved forward quickly and planned to press that in this week's meeting.

This PR is just waiting on feedback from either @artlowel or @atarix83 .  Since it's a larger sized PR, I'd like to get two approvals at a minimum (per our dev policies ), and so far it just has @wwelling's approval.

File: src/modules/app/browser-app.module.ts
Comment (artlowel) [inline src/modules/app/browser-app.module.ts:62]: Did you check to see whether `relativeLinkResolution: 'legacy'` is needed? At first sight I couldn't find any instances of [the bug it's trying to stay backwards compatible with](https://angular.io/api/router/ExtraOptions#relativeLinkResolution)

**Review by artlowel — APPROVED** [review]
Thanks @tdonohue!

I tested a few dynamic themes, and I found no issues

File: src/modules/app/browser-app.module.ts
Comment (tdonohue) [reply src/modules/app/browser-app.module.ts:62]: @artlowel : this was an *automated change*. It looks like it was automatically added when I ran `npx @angular/cli@11 update @angular/core@11 @angular/cli@11 --force`.  If we wanted to try to remove it, we could do so (either in this PR or in a later PR). I avoided making any manual changes in this PR except where necessary.

My assumption though is that the `update` script must have seen something in our code that implied it needed to enable `legacy` behavior.  Either that, or it's enabled by default in every Angular 11 upgrade.

**Comment by atarix83** [comment]
@tdonohue @wwelling 

The following dependencies can be updateted too as indicated : 

"@angular-builders/custom-webpack": "11.1.1" 
"@ngtools/webpack": "11.2.15"
"karma-jasmine-html-reporter": "~1.5.4"

File: src/modules/app/browser-app.module.ts
Comment (atarix83) [reply src/modules/app/browser-app.module.ts:62]: @artlowel @tdonohue 

for what says the documentation here https://v11.angular.io/api/router/ExtraOptions#relativeLinkResolution if we don't use the legacy option we should check if all relative links work properly

**Comment by tdonohue** [comment]
@atarix83 : I attempted an upgrade of custom-webpack and webpack by doing this...
```
ng update @ngtools/webpack@11 @angular-builders/custom-webpack@11 --force
```

Unfortunately though, that requires us to also upgrade every single PostCSS dependency, as it creates a dependency on PostCSS8.  So, I tried that..

```
ng update postcss-cli@8 postcss-preset-env@7 postcss-loader@4 css-loader@4 cssnano@5 postcss-import@12
```

However, that in turn resulted in very odd PostCSS related build failures when running `yarn start`.  I've not been able to track down the root cause of the issues, and I cannot get the application to build after upgrading to PostCSS v8.

So, I'd recommend we consider treating this as a separate ticket, since the update to PostCSS v8 is not simple.

File: src/modules/app/browser-app.module.ts
Comment (tdonohue) [reply src/modules/app/browser-app.module.ts:62]: As discussed in today's meeting, I'll try and remove this `relativeLinkResolution: 'legacy'` option today and retest. If it works fine, I'll push it up into this PR & get a final review before we merge this (as we agreed we want to merge this soon). If it doesn't work (and I cannot easily figure out a solution), I'll log a separate ticket to remove the `relativeLinkResolution: 'legacy'` & we'll move this PR forward as-is.

Comment (tdonohue) [reply src/modules/app/browser-app.module.ts]: @artlowel and @atarix83 : I removed `relativeLinkResolution: 'legacy'` and manually checked our links to make sure we didn't need it (I couldn't find anything that required legacy behavior).  After removing it, I retested everything using `yarn start` and I'm seeing no change in behavior.  

Could you retest?  That way we can get this merged as soon as reasonably possible.

**Review by artlowel — APPROVED** [review]
I can't find any links that are broken

**Comment by wwelling** [comment]
@atarix83 would you be willing to sign off on this PR? I do not think any more dependencies need to be upgraded here.

**Comment by tdonohue** [comment]
@wwelling : @artlowel , @atarix83 and I talked about this in yesterday's Dev Meeting, and decided that this PR can be merged as soon as two of us test out the final change in ad5a76a.  Since @artlowel and I have done that, I'm merging this PR immediately.

I will create a follow-up ticket shortly around @atarix83 's [feedback](https://github.com/DSpace/dspace-angular/pull/1403#issuecomment-984668080)... I do agree with him that we should work to upgrade those other dependencies, but that upgrade seems reliant on [a more complex PostCSS v8 upgrade](https://github.com/DSpace/dspace-angular/pull/1403#issuecomment-984709353).  So, we decided to split that task out into a separate ticket, which would allow the basic Angular 11 upgrade to be merged immediately.  I'll create that separate ticket, and schedule it for 7.2...hopefully one of us will find some time to complete it during the 7.2 release (I don't have any immediate time myself, unfortunately)

**Pull request merged**
